Expose torch._will_engine_execute_node

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84773

Addresses: https://github.com/pytorch/pytorch/issues/83617

This PR a way to query the TLS graph task's exec_info which is a map mapping the Node to a bool indicating whether it will be executed in the current backward pass (as determined by the inputs= argument for .grad of .backward).
- this works with both custom Function nodes and normal codegened nodes
-  to be able to verify whether the pyobject passed is an actual node, we now store pointers to PyTypeObjects into a set on registration.
- error out when .backward without inputs= to avoid silently returning True

Alternatives:
- not sure if it is possible to bind to Python from a raw pointer to Node. At least we wouldn't be able to use existing logic, and the Python object should only hold a weak reference to the Node.
- other solutions to the motivating issue seem to require more extensive modification to the engine 

See the issue linked for an example of usage

soulitzer committed:
Expose torch._will_engine_execute_node

[ghstack-poisoned]

soulitzer committed:
Update on "Expose torch._will_engine_execute_node"


Modified Alban's example to use new API:

```python
import torch

def register_multi_grad_hook(tensors, fn):
    count = 0
    nb_calls = None
    buffer = None

    def get_grad_fn(t):
        # or grad accumulator
        if t.requires_grad and t.grad_fn is None:
            return t.clone().grad_fn.next_functions[0][0]
        else:
            return t.grad_fn

    grad_fns = list(map(get_grad_fn, tensors))

    def get_inner_hook(idx):
        def inner_hook(grad):
            nonlocal count, nb_calls, buffer

            if count == 0:
                # On the first call, compute the actual nb_calls and buffer
                nb_calls = sum(1 for g in grad_fns if torch._C._will_engine_execute_node(g))
                buffer = [None] * nb_calls

            buffer[idx] = grad
            count += 1

            if count == nb_calls:
                fn(buffer)
        return inner_hook
    for i, t in enumerate(tensors):
        t.register_hook(get_inner_hook(i))


t1 = torch.rand(2, requires_grad=True)
t2 = torch.rand(2, requires_grad=True)
t3 = torch.rand(2, requires_grad=True)
t4 = torch.rand(2, requires_grad=True)

def hook(grads):
    print(f"Multi-hook called with {len(grads)} gradients")
register_multi_grad_hook((t2, t3), hook)

def get_hook(name):
    def hook(grad):
        print(f"{name} hook called")
    return hook
t1.register_hook(get_hook("t1"))
t3.register_hook(get_hook("t3"))


out = t1.clone()
out = out + t2
out = out + t3
out = out + t4

out.sum().backward(inputs=(t2, t3))
# t3 hook called
# Multi-hook called with 2 gradients
# t1 hook called

out.sum().backward(inputs=(t1, t2))
# Multi-hook called with 1 gradients
# t1 hook called
```

[ghstack-poisoned]

soulitzer committed:
Update on "Expose torch._will_engine_execute_node"


Modified Alban's example to use new API:

```python
import torch

def register_multi_grad_hook(tensors, fn):
    count = 0
    nb_calls = None
    buffer = None

    def get_grad_fn(t):
        # or grad accumulator
        if t.requires_grad and t.grad_fn is None:
            return t.clone().grad_fn.next_functions[0][0]
        else:
            return t.grad_fn

    grad_fns = list(map(get_grad_fn, tensors))

    def get_inner_hook(idx):
        def inner_hook(grad):
            nonlocal count, nb_calls, buffer

            if count == 0:
                # On the first call, compute the actual nb_calls and buffer
                nb_calls = sum(1 for g in grad_fns if torch._C._will_engine_execute_node(g))
                buffer = [None] * nb_calls

            buffer[idx] = grad
            count += 1

            if count == nb_calls:
                fn(buffer)
        return inner_hook
    for i, t in enumerate(tensors):
        t.register_hook(get_inner_hook(i))


t1 = torch.rand(2, requires_grad=True)
t2 = torch.rand(2, requires_grad=True)
t3 = torch.rand(2, requires_grad=True)
t4 = torch.rand(2, requires_grad=True)

def hook(grads):
    print(f"Multi-hook called with {len(grads)} gradients")
register_multi_grad_hook((t2, t3), hook)

def get_hook(name):
    def hook(grad):
        print(f"{name} hook called")
    return hook
t1.register_hook(get_hook("t1"))
t3.register_hook(get_hook("t3"))


out = t1.clone()
out = out + t2
out = out + t3
out = out + t4

out.sum().backward(inputs=(t2, t3))
# t3 hook called
# Multi-hook called with 2 gradients
# t1 hook called

out.sum().backward(inputs=(t1, t2))
# Multi-hook called with 1 gradients
# t1 hook called
```

[ghstack-poisoned]

soulitzer committed:
Update on "Expose torch._will_engine_execute_node"


Modified Alban's example to use new API:

```python
import torch

def register_multi_grad_hook(tensors, fn):
    count = 0
    nb_calls = None
    buffer = None

    def get_grad_fn(t):
        # or grad accumulator
        if t.requires_grad and t.grad_fn is None:
            return t.clone().grad_fn.next_functions[0][0]
        else:
            return t.grad_fn

    grad_fns = list(map(get_grad_fn, tensors))

    def get_inner_hook(idx):
        def inner_hook(grad):
            nonlocal count, nb_calls, buffer

            if count == 0:
                # On the first call, compute the actual nb_calls and buffer
                nb_calls = sum(1 for g in grad_fns if torch._C._will_engine_execute_node(g))
                buffer = [None] * nb_calls

            buffer[idx] = grad
            count += 1

            if count == nb_calls:
                fn(buffer)
        return inner_hook
    for i, t in enumerate(tensors):
        t.register_hook(get_inner_hook(i))


t1 = torch.rand(2, requires_grad=True)
t2 = torch.rand(2, requires_grad=True)
t3 = torch.rand(2, requires_grad=True)
t4 = torch.rand(2, requires_grad=True)

def hook(grads):
    print(f"Multi-hook called with {len(grads)} gradients")
register_multi_grad_hook((t2, t3), hook)

def get_hook(name):
    def hook(grad):
        print(f"{name} hook called")
    return hook
t1.register_hook(get_hook("t1"))
t3.register_hook(get_hook("t3"))


out = t1.clone()
out = out + t2
out = out + t3
out = out + t4

out.sum().backward(inputs=(t2, t3))
# t3 hook called
# Multi-hook called with 2 gradients
# t1 hook called

out.sum().backward(inputs=(t1, t2))
# Multi-hook called with 1 gradients
# t1 hook called
```

[ghstack-poisoned]

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -688,6 +690,38 @@ PyObject* THPModule_isEnabledXNNPACK(PyObject* _unused, PyObject* noargs) {
+      "if the `inputs` is passed.");
+  torch::autograd::Node* node;
+  if (isTHPFunction) {
+    node = ((THPFunction*)arg)->cdata.lock().get();
```
Is this OK? Aren't you obligated to keep lock's output live?

soulitzer replied:
Good catch, updated

soulitzer replied:
actually since we don't dereference the ptr it is probably OK, though it makes sense in principle

ezyang replied:
Keeping it live prevents ABA races, although since you take out a mutex that might also be sufficient protection

soulitzer committed:
Update on "Expose torch._will_engine_execute_node"


Addresses: https://github.com/pytorch/pytorch/issues/83617

This PR a way to query the TLS graph task's exec_info which is a map mapping the Node to a bool indicating whether it will be executed in the current backward pass (as determined by the inputs= argument for .grad of .backward).
- this works with both custom Function nodes and normal codegened nodes
-  to be able to verify whether the pyobject passed is an actual node, we now store pointers to PyTypeObjects into a set on registration.
- error out when .backward without inputs= to avoid silently returning True

Alternatives:
- not sure if it is possible to bind to Python from a raw pointer to Node. At least we wouldn't be able to use existing logic, and the Python object should only hold a weak reference to the Node.
- other solutions to the motivating issue seem to require more extensive modification to the engine 

See the issue linked for an example of usage

[ghstack-poisoned]

ezyang commented:
This looks reasonable, though I guess maybe we should wait for albanD to check this

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -688,6 +690,40 @@ PyObject* THPModule_isEnabledXNNPACK(PyObject* _unused, PyObject* noargs) {
+      "_get_should_execute_nodes should only be called during the backward pass");
+  THPUtils_assert(
+      !exec_info->empty(),
+      "_get_should_execute_nodes should only be used with .grad() or .backward() "
```
Why is that?
Even in the case of `backward()`, we need to know how many gradients will be computed for this function and if that number is >1 that means this function will be executed.

soulitzer replied:
We don't want to erroneously return True when there is a Tensor that is not part the graph, so we'd have to know what nodes will be traversed, when we eventually traverse the graph (I'm not sure if we have this information, but I can double check). If it's not already there, adding it seems to be additional overhead that we might want to avoid.

albanD replied:
If the Tensor is not part of the graph, then this Node will have 0 required gradients (or more likely no entry in the required gradients table). Meaning that no gradient will be computed for it.
This is stored in the `dependencies_` field of the graph task btw.

soulitzer replied:
Thanks for the pointer, updated!

albanD reviewed (commented):
The approach sounds good, not sure why the limitation on backward() though

soulitzer committed:
Update on "Expose torch._will_engine_execute_node"


Addresses: https://github.com/pytorch/pytorch/issues/83617

This PR a way to query the TLS graph task's exec_info which is a map mapping the Node to a bool indicating whether it will be executed in the current backward pass (as determined by the inputs= argument for .grad of .backward).
- this works with both custom Function nodes and normal codegened nodes
-  to be able to verify whether the pyobject passed is an actual node, we now store pointers to PyTypeObjects into a set on registration.
- error out when .backward without inputs= to avoid silently returning True

Alternatives:
- not sure if it is possible to bind to Python from a raw pointer to Node. At least we wouldn't be able to use existing logic, and the Python object should only hold a weak reference to the Node.
- other solutions to the motivating issue seem to require more extensive modification to the engine 

See the issue linked for an example of usage

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/graph_task.h:
```diff
@@ -31,6 +31,8 @@ struct GraphTask : std::enable_shared_from_this<GraphTask> {
   std::unordered_map<Node*, int> dependencies_;
 
+  // Records the nodes that are in the graph
+  std::unordered_set<Node*> nodes_in_graph_;
```
To access `dependencies_` we'd need to acquire the graph task mutex. That might be problematic because we need to acquire the same mutex when writing to graph_task->captured_vars_ and call user-defined capture hooks. If the user-defined code in those capture hooks calls into our new API that would lead to a deadlock.

soulitzer replied:
Reading dependencies_ requires the mutex to be acquired, and someone might try to call this API when the mutex is already acquired, e.g. when capture variable hooks are called.

albanD replied:
We never add new elements to the map though, we only change some values. Is it still unsafe to check if a key is in the map in that context?

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -688,6 +690,50 @@ PyObject* THPModule_isEnabledXNNPACK(PyObject* _unused, PyObject* noargs) {
+    } else {
+      Py_RETURN_TRUE;
+    }
+  } else {
```
Why do we need this case? The new field covers this case as well.

soulitzer replied:
We should need both because compute_dependencies just traverses through the graph without taking into account the inputs= argument.

albanD replied:
No it does not? Otherwise the `dependencies_` counter would be wrong?

soulitzer replied:
From reading the code again, what happens later is that we want to use dependencies_ to see if a given node is ready to be executed, but first we use exec_info to decide whether we want to execute the node in the first place, so it is indeed possible that a node will have an entry in dependencies_, but won't actually be executed.

albanD replied:
Ho right at the border of the of the graph there is one more node for which the hook won't be called?
But is that the expected behavior btw? (there is an issue for that somewhere right?)

soulitzer replied:
Ohh that is a very good catch, neither variables hooks registered to leafs nor function hooks registered to acc grad are called when .grad is used. 

As you can see from the test, exec_info_ does not know about that either though. 

I think we should error out in this case, as opposed to silently returning false for accumulate grad nodes. If someone passes in an acc grad node to our function when running .grad that is probably user error.

albanD reviewed (approved):
Ok! This is consistent with the current hook behavior.

I think it is time to pay the cost and actually do the BC-breaking change of enabling hook on inputs= as a next step. WDYT?

soulitzer committed:
Update on "Expose torch._will_engine_execute_node"


Addresses: https://github.com/pytorch/pytorch/issues/83617

This PR a way to query the TLS graph task's exec_info which is a map mapping the Node to a bool indicating whether it will be executed in the current backward pass (as determined by the inputs= argument for .grad of .backward).
- this works with both custom Function nodes and normal codegened nodes
-  to be able to verify whether the pyobject passed is an actual node, we now store pointers to PyTypeObjects into a set on registration.
- error out when .backward without inputs= to avoid silently returning True

Alternatives:
- not sure if it is possible to bind to Python from a raw pointer to Node. At least we wouldn't be able to use existing logic, and the Python object should only hold a weak reference to the Node.
- other solutions to the motivating issue seem to require more extensive modification to the engine 

See the issue linked for an example of usage

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

soulitzer commented:
@pytorchbot merge -f "Unrelated failure"
